java-components/cache/src/main/java/com/redhat/hacbs/artifactcache/services/client/ociregistry/OCIRegistryRepositoryClient.java

```diff
@@ -127,53 +136,46 @@ private RegistryClient getRegistryClient(String registry, String owner, String g
             throws IOException {
-        if (imageCache.containsKey(digestHash)) {
-            return pullStale(registryClient, manifest, digestHash);
-        } else {
-            return pullFreshAndCache(registryClient, manifest, digestHash);
-        }
-    }
-
-    private Path pullStale(RegistryClient registryClient, ManifestTemplate manifest, String digestHash) throws IOException {
-        Path path = imageCache.get(digestHash);
-        if (Files.exists(path)) {
-            return path;
+        Path digestHashPath = cacheRoot.resolve(digestHash);
+        if (existInLocalCache(digestHashPath)) {
+            return Optional.of(Paths.get(digestHashPath.toString(), ARTIFACTS));
         } else {
-            return pullFreshAndCache(registryClient, manifest, digestHash);
+            return pullFromRemoteAndCache(registryClient, manifest, digestHash, digestHashPath);
         }
-
     }
 
-    private Path pullFreshAndCache(RegistryClient registryClient, ManifestTemplate manifest, String digestHash)
+    private Optional<Path> pullFromRemoteAndCache(RegistryClient registryClient, ManifestTemplate manifest, String digestHash,
+            Path digestHashPath)
             throws IOException {
 
         String manifestMediaType = manifest.getManifestMediaType();
 
         if (OCI_MEDIA_TYPE.equalsIgnoreCase(manifestMediaType)) {
             List<BuildableManifestTemplate.ContentDescriptorTemplate> layers = ((OciManifestTemplate) manifest).getLayers();
+            if (layers.size() == 3) {
+                // Layer 2 is artifacts
+                BuildableManifestTemplate.ContentDescriptorTemplate artifactsLayer = layers.get(2);
 
-            // Layer 2 is artifacts
-            BuildableManifestTemplate.ContentDescriptorTemplate artifactsLayer = layers.get(2);
-
-            Blob blob = registryClient.pullBlob(artifactsLayer.getDigest(), s -> {
-            }, s -> {
-            });
+                Blob blob = registryClient.pullBlob(artifactsLayer.getDigest(), s -> {
+                }, s -> {
+                });
 
-            ByteArrayOutputStream out = new ByteArrayOutputStream();
-            blob.writeTo(out);
+                Path outputPath = Files.createDirectories(digestHashPath);
 
-            // Extract to temp (TODO: Configure this ?)
-            String osTmpDir = System.getProperty(SYS_PROP_TEMP_DIR);
-            Path tempFullPath = Paths.get(osTmpDir, HACBS);
-            Path outputPath = Files.createDirectories(tempFullPath);
-            extractTarArchive(out.toByteArray(), outputPath.toString());
-            Path repoRoot = Paths.get(outputPath.toString(), ARTIFACTS);
-
-            imageCache.put(digestHash, repoRoot);
-
-            return repoRoot;
+                Path tarFile = Files.createFile(Paths.get(outputPath.toString(), digestHash + ".tar"));
+                try (OutputStream tarOutputStream = Files.newOutputStream(tarFile)) {
+                    blob.writeTo(Files.newOutputStream(tarFile));
```

Comment:
easy enough / cool :-)

Reply:
oops, I missed that in review, @phillip-kruger shouldn't think be using the tarOutputStream ?